src/bin/ion/commands/generate/generator.rs

```diff
@@ -501,6 +509,76 @@ impl<'a, L: Language + 'static> CodeGenerator<'a, L> {
 
         Ok(AbstractDataType::Structure(structure_builder.build()?))
     }
+
+    /// Build wrapped scalar data type from constraints
+    fn build_wrapped_scalar_from_constraints(
+        &mut self,
+        constraints: &[IslConstraint],
+        code_gen_context: &mut CodeGenContext,
+        parent_isl_type: &IslType,
+    ) -> CodeGenResult<AbstractDataType> {
+        let mut wrapped_scalar_builder = WrappedScalarBuilder::default();
+        wrapped_scalar_builder
+            .name(self.current_type_fully_qualified_name.to_owned())
+            .source(parent_isl_type.to_owned());
+        for constraint in constraints {
+            match constraint.constraint() {
+                IslConstraintValue::Type(isl_type) => {
+                    let type_name = self
+                        .fully_qualified_type_ref_name(isl_type, code_gen_context)?
+                        .ok_or(invalid_abstract_data_type_raw_error(format!(
+                            "Could not determine `FullQualifiedTypeReference` for type {:?}",
+                            isl_type
+                        )))?;
+
+                    // by default fields aren't closed
+                    wrapped_scalar_builder.base_type(type_name);
+                }
+                _ => {
+                    return invalid_abstract_data_type_error(
+                        "Could not determine the abstract data type due to conflicting constraints",
+                    );
+                }
+            }
+        }
+
+        Ok(AbstractDataType::WrappedScalar(
+            wrapped_scalar_builder.build()?,
+        ))
+    }
+
+    /// Build scalar data type from constraints
+    fn build_scalar_from_constraints(
+        &mut self,
+        constraints: &[IslConstraint],
+        code_gen_context: &mut CodeGenContext,
+        parent_isl_type: &IslType,
+    ) -> CodeGenResult<AbstractDataType> {
+        let mut scalar_builder = ScalarBuilder::default();
+        scalar_builder.source(parent_isl_type.to_owned());
+        for constraint in constraints {
+            match constraint.constraint() {
+                IslConstraintValue::Type(isl_type) => {
+                    let type_name = self
+                        .fully_qualified_type_ref_name(isl_type, code_gen_context)?
+                        .ok_or(invalid_abstract_data_type_raw_error(format!(
+                            "Could not determine `FullQualifiedTypeReference` for type {:?}",
```

Comment:
If this happens, is it due to invalid input or a bug?

Reply:
It will only be `None` in the case of `struct`, `list`, `sexp` as we can not determine a target type for these open ended container types. But I think making this error message say that would make things clear.